### DISCUSSION TITLE
Data logging

### DISCUSSION THREAD

**Discussion by DeeEmm** [DIY-Flow-Bench/DIY-Flow-Bench#184] [Code Development]
At the moment the system uses a very basic JSON file to record flow values across the lift range and displays the points on the lift graph. This is of course of basic value and does not include additional useful info such as environmental conditions reference pressure etc.

Current JSON file...

<img width="165" alt="Screenshot 2024-10-29 at 11 48 23 am" src="https://github.com/user-attachments/assets/d2697a40-a1b7-4daa-b243-ae9d69ed207a">

The plan was always to expand the types of data recorded and displayed on the lift graph and also to be able to load up any previously recorded data to view it in the GUI.

It makes sense to simply record all data that we have available...

<img width="920" alt="Screenshot 2024-10-29 at 11 50 35 am" src="https://github.com/user-attachments/assets/3020096b-5367-4704-a274-101a09ee8d4d">

This data can be recorded for each data point.

Further to this the ability to load more than one record simultaneously on the data graph allows for direct comparison.

Whilst these features are on the roadmap, they are not in current development. But as they have been mentioned in discussion it seems like a good time to create a discussion thread about it and try to determine the best approach.

Thoughts...?

**Comment by stefan63** [comment]
Loggging to google sheets, however that will require a good internet connection and we might not want to relay on that.
Having Node-Red and influx, Grafana runing on the local network is another aproach to get infinite data logging capability and you could also build the UI in Node-red.
https://community.element14.com/products/arduino/b/blog/posts/sending-data-from-esp32-arduino-to-node-red-and-storing-to-influxdb-using-websocket

**Reply by DeeEmm** [reply]
Node-Red is an interesting concept. We already employ both websockets and JSON in the current code.  Is there a use-case for streaming data in this fashion?

I will admit that I always assumed data collection would be handled on the device and could then be exported in JSON format via the browser for external consumption if required, but I guess that all comes down to individual use cases. My experience on relying on external integrations is that it makes the project a lot more work to manage. 

Also, the SD card integration should also be considered as well. Whilst this is not yet implemented, it is another method of storage outside of SPIFFs. The ESP can support up to 2TB SD which should provide many lifetimes of flow jobs.

and on that....

It made me think of accessing the ESP as a network share over SMB / SAMBA which I think is doable.

**Reply by stefan63** [reply]
I agree that using external server will complicate things and make it less acessable for the less techy users.
That made me thinking that this might have been better to deploy on a Raspbery Pi then you can distribute a complete ISO and just dress the Pi with a nice suitable hat. The pi 5 does now have official suport for HDD's and you can hook up an old screen mouse and keyboard. No wifi or other ui device neded .  I understand that this will require a lot of code rework but it could be a future project. For now logging to SD card and be able to export it to a spreadsheet will work just fine and also makes it possible to customize the diagrams to your personal taste.

**Reply by stefan63** [reply]
For SAMBA have a look at this.
https://github.com/nopnop2002/esp-idf-smb-client
And
https://github.com/jsebgiraldo/ESP32-SMB-API
However the last link doesnt mention put mode for SMB while the first one does.

**Reply by DeeEmm** [reply]
Ha. No. The ESP32 is the right tool for the job. 

Pi's are microprocessors whereas ESP32's are microcontrollers. 

Use a PI if you want to do computing
Use an arduino / esp32 if you want to control I/O as they are much better at it. It's literally the one job they are designed to do.

Remember the primary job of the system is to measure and convert flow. Which the ESP will do far more reliably than the PI. 

IMHO using a PI for managing data logging is overkill. We are only talking about a few data fields for each recorded data point. There's no streaming data or large data sets. All of the data from a complete test run across 12 lift points fits inside a 2kb file. There's no need for a microprocessor as we don't actually need to process anything.

**Reply by stefan63** [reply]
I agree.

**Comment by black-top** [comment]
Is there any simple way to implement " record all data that we have available " ?
I think importing data from current lifdata.jason to excel spreadsheet for data analysis and graphical display will be good enough  until more sophisticated system will be implemented in the GUI.

**Reply by DeeEmm** [reply]
> Is there any simple way to implement " record all data that we have available " ? I think importing data from current lifdata.jason to excel spreadsheet for data analysis and graphical display will be good enough until more sophisticated system will be implemented in the GUI.

Yes. in the API, if you issue the command 'E' (note capitalisation) it will reply with an enum that has the following format:

Enum - Flow:Ref:Temp:Humidity:Baro

Whilst this does not save the data on the device, you can actually access this directly from excel.

You can set Excel to use the serial port as a data source and grab data from it. It involves a little bit of VBA to get the data exchange working. (The following code might serve as a starting point - https://strokescribe.com/en/serial-port-excel.html )

Basically every time excel sends "E" to the DIYFB serial port, it will respond with the enum above.

**Comment by black-top** [comment]
With the current flow bench setup and small air pump, I get about 200CFM @28" and, at about 300CFM,  pressure drops to 12-14"h2o. So uncorrected value plateaued at about 200CFM and corrected value keeps going to 300+CFM, and it is quite similar to what is expected of this head. 
The corrected values I was writing by hand, so it will be good to implement "capture all data" for more accurate comparisons.

Soon I'm planning to put together another similar flow bench with a much bigger pump, and it should flow at least 500CFM @ 28" , so it will be interesting to compare results corrected and uncorrected with a small and big pump flow benches.

**Comment by DeeEmm** [comment]
So I tried to see if I could get the import into Excel working, but I am on Mac and the required 'Data Streamer' plugin is not available for Excel for Mac. It should however work for those of you on Windows

https://learn.microsoft.com/en-us/microsoft-365/education/data-streamer/

I do however use a serial terminal emulator called CoolTerm fairly often so decided to take a look at that instead. It is trivial to set up. Make sure that you do not already have a serial connection in VScode and set the baud rate

<img width="518" alt="Screenshot 2024-11-04 at 5 34 57 pm" src="https://github.com/user-attachments/assets/292b609a-74d4-42b1-8e40-a65f91ef0c13">

Then just hit connect.

You can then send any of the API commands to the serial monitor...

<img width="1096" alt="Screenshot 2024-11-04 at 5 36 34 pm" src="https://github.com/user-attachments/assets/39de464e-63ae-4ec9-8bc3-3ed28fee7a55">

Here's the bonus... CoolTerm also has the ability to capture to file...

<img width="649" alt="Screenshot 2024-11-04 at 5 37 26 pm" src="https://github.com/user-attachments/assets/f465f442-fe7c-4605-a5a0-5ac11e3717ba">

Just select where you want to save the data

<img width="1924" alt="Screenshot 2024-11-04 at 5 39 20 pm" src="https://github.com/user-attachments/assets/b6f24625-72ed-479e-8699-2754d61f4fa3">

I previously forgot there is also a JSON version of the Status Data (as can be seen in the screen shot).

We can easily tailor a response to include whatever data we want. However the JSON data is everything that is sent to the GUI

I've added the JSON call to HTTP API requests as well...

<img width="1491" alt="Screenshot 2024-11-04 at 5 57 41 pm" src="https://github.com/user-attachments/assets/9cee142b-aa70-4806-a82e-82a360f9689e">

```
  // Send JSON Data
  server->on("/api/json", HTTP_GET, [](AsyncWebServerRequest *request){
    Webserver _webserver;
    request->send(200, "text/html", String(_webserver.getDataJSON()).c_str());
  });
```

I will create a PR with this update.

You can access it via http://diyfb.local/api/json

We should also be able to pull data in from Excel with this but it would require a small bit of VBA code to send the necessary "J" char to the terminal

So a few methods to get data out of DIYFB, both tethered and OTA.

I want to build on this format to save data for the data validation tests (#35) but append to a file in SPIFFS.

**Comment by stefan63** [comment]
Self hosted "Firebase" Appwrite
This can be a alternative data logging storage solution.
Thinking Appwriter running on a PI Zero and a USB disk or your pc used as controll device for Flow bench UI.
https://github.com/appwrite/appwrite

**Reply by DeeEmm** [reply]
Yes. Any external app is possible if it can access the API - either by serial or via http, so something like a Pi or a computer is perfect.

There are a number of solutions that produce nice graphical representations of data. I did consider integrating a similar library in the past for viewing data in the UI, which may still happen at some point after the main code is done. The only issue with libraries such as that is that they usually take up a lot of space. 

There is also the use of HTML5 which can produce awesome results, which we can directly integrate without needing a bloaty library. Lots of options.
